File: 5212R/utils/swatch/incus_load.go
// incus_load.go
package main

import (
    "encoding/json"
    "fmt"
    "log"
    "os"
    "os/exec"
    "path/filepath"
    "strconv"
    "strings"
    "time"
)

type Container struct {
    Name   string `json:"name"`
    Status string `json:"status"`
    State  struct {
        PID int `json:"pid"`
    } `json:"state"`
}

func getUptime() string {
    out, err := exec.Command("uptime").Output()
    if err != nil {
        return "N/A"
    }
    return strings.TrimSpace(string(out))
}

func parseProcUptime() (int, int, int) {
    data, err := os.ReadFile("/proc/uptime")
    if err != nil {
        return 0, 0, 0
    }
    parts := strings.Fields(string(data))
    sec, _ := strconv.ParseFloat(parts[0], 64)
    days := int(sec) / 86400
    hours := int(sec)%86400 / 3600
    minutes := int(sec)%3600 / 60
    return days, hours, minutes
}

func formatTime(days, hours, minutes int) string {
    return fmt.Sprintf("%dd:%02dh:%02dmin", days, hours, minutes)
}

func loadContainers() ([]Container, error) {
    out, err := exec.Command("/usr/bin/incus", "list", "--format=json").Output()
    if err != nil {
        return nil, err
    }
    var containers []Container
    err = json.Unmarshal(out, &containers)
    return containers, err
}

func printRunningContainers(containers []Container, maxWidth int) {
    // Adjust column widths for better alignment
    fmt.Printf("%-*s%-18s%-12s%-16s%-16s\n", maxWidth, "CONTAINER", "UPTIME", "TIME", "USERS", "LOAD AVERAGE")

    for _, c := range containers {
        if c.Status != "Running" {
            continue
        }

        // Get uptime in seconds from /proc/uptime inside the container
        upCmd := exec.Command("/usr/bin/incus", "exec", c.Name, "--", "cat", "/proc/uptime")
        upRaw, _ := upCmd.Output()
        days, hours, minutes := 0, 0, 0
        if fields := strings.Fields(string(upRaw)); len(fields) > 0 {
            if sec, err := strconv.ParseFloat(fields[0], 64); err == nil {
                days = int(sec) / 86400
                hours = int(sec)%86400 / 3600
                minutes = int(sec)%3600 / 60
            }
        }
        uptimeFmt := formatTime(days, hours, minutes)

        // Get uptime and load average with users from the 'uptime' command inside the container
        uptimeCmd := exec.Command("/usr/bin/incus", "exec", c.Name, "--", "uptime")
        uptimeOut, _ := uptimeCmd.Output()
        uptimeStr := string(uptimeOut)

        // Initialize user and load values
        users := "0" // Default to 0 users
        load := "N/A" // Default load average

        // Parse uptimeStr to extract the correct users count and load averages
        if strings.Contains(uptimeStr, "load average") {
            // Extract load average part
            parts := strings.Split(uptimeStr, "load average:")
            load = strings.TrimSpace(parts[1])

            // Extract user count from uptime string (e.g., "0 user" or "1 user")
            userParts := strings.Split(uptimeStr, ", ")
            if len(userParts) > 2 {
                users = strings.TrimSpace(userParts[2]) // Correctly extract the "X user" part
                users = strings.Split(users, " ")[0]    // Remove "user" or "users" from the string
            } else {
                users = "0"  // Default users to 0 if not found
            }
        }

        // Format userStr correctly (i.e., "1 user" or "X users")
        userStr := fmt.Sprintf("%s user", users)
        if users != "1" {
            userStr = fmt.Sprintf("%s users", users)
        }

        // Adjust column widths for better alignment in the print statement
        fmt.Printf("%-*s%-18s%-12s%-16s%-16s\n", maxWidth, c.Name, uptimeFmt, time.Now().Format("15:04:05"), userStr, load)
    }
}

func listProcesses(containerMap map[string]string, maxWidth int) {
    fmt.Printf("%-10s%-6s%-6s%-23s%-*s\n", "PID", "CPU%", "MEM%", "COMMAND", maxWidth, "CONTAINER")

    out, err := exec.Command("ps", "-e", "h", "-o", "pid,pcpu,pmem,comm", "--sort=-pcpu").Output()
    if err != nil {
        return
    }
    lines := strings.Split(string(out), "\n")
    for _, line := range lines {
        fields := strings.Fields(line)
        if len(fields) < 4 {
            continue
        }
        pid := fields[0]
        cpu, _ := strconv.ParseFloat(fields[1], 64)
        if cpu <= 0.3 {
            continue
        }
        mem := fields[2]
        command := fields[3]
        container := "host"
        procNs := filepath.Join("/proc", pid, "ns/pid")
        if procLink, err := os.Readlink(procNs); err == nil {
            for hpid, cname := range containerMap {
                hostNs := filepath.Join("/proc", hpid, "ns/pid")
                if hostLink, err := os.Readlink(hostNs); err == nil && procLink == hostLink {
                    container = cname
                    break
                }
            }
        }
        fmt.Printf("%-10s%-6s%-6s%-23s%-*s\n", pid, fields[1], mem, command, maxWidth, container)
    }
}

func main() {
    if _, err := os.Stat("/usr/bin/incus"); err != nil {
        fmt.Println("Incus is not installed. Not reporting Instance load averages.")
        return
    }

    uptime := getUptime()
    node, _ := os.Hostname()
    fmt.Println("NODE:")
    fmt.Println(node, uptime)
    fmt.Println()

    containers, err := loadContainers()
    if err != nil {
        log.Fatal("Failed to parse Incus container list")
    }

    nameWidths := 15
    names := []string{}
    containerPidMap := map[string]string{}
    for _, c := range containers {
        if c.Status == "Running" {
            names = append(names, c.Name)
            containerPidMap[strconv.Itoa(c.State.PID)] = c.Name
        }
    }
    for _, name := range names {
        if len(name) > nameWidths {
            nameWidths = len(name) + 1
        }
    }

    printRunningContainers(containers, nameWidths)
    fmt.Println()
    listProcesses(containerPidMap, nameWidths)
}

//
// Copyright (c) 2008-2025 Michael Stauber,  SOLARSPEED.NET
// Copyright (c) 2008-2025 Greg Kuhnert,     COMPASSNETWORKS.COM.AU
// Copyright (c) 2008-2025 Team BlueOnyx,    BLUEONYX.IT
// All Rights Reserved.
// 
// 1. Redistributions of source code must retain the above copyright 
//    notice, this list of conditions and the following disclaimer.
// 
// 2. Redistributions in binary form must reproduce the above copyright 
//    notice, this list of conditions and the following disclaimer in 
//    the documentation and/or other materials provided with the 
//    distribution.
// 
// 3. Neither the name of the copyright holder nor the names of its 
//    contributors may be used to endorse or promote products derived 
//    from this software without specific prior written permission.
// 
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
// COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN 
// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
// POSSIBILITY OF SUCH DAMAGE.
// 
// You acknowledge that this software is not designed or intended for 
// use in the design, construction, operation or maintenance of any 
// nuclear facility.
//

File: 5212R/utils/swatch/incus_load.py
import os
import subprocess
import json
import time

def get_uptime():
    """Get the system uptime."""
    result = subprocess.run(["uptime"], capture_output=True, text=True)
    return result.stdout.strip()

def parse_uptime_proc():
    """Parse uptime from /proc/uptime."""
    try:
        with open("/proc/uptime", "r") as f:
            uptime_seconds = float(f.readline().split()[0])
            days = int(uptime_seconds // 86400)
            hours = int((uptime_seconds % 86400) // 3600)
            minutes = int((uptime_seconds % 3600) // 60)
            return days, hours, minutes
    except Exception as e:
        print(f"Error reading /proc/uptime: {e}")
        return 0, 0, 0

def format_time(days, hours, minutes):
    """Format uptime with leading zeros."""
    return f"{days}d:{hours:02}h:{minutes:02}min"

def map_processes_to_containers(max_container_width):
    """Map processes to containers and display their information with dynamic width."""
    print(f"{'PID':<10}{'CPU%':<6}{'MEM%':<6}{'COMMAND':<23}{'CONTAINER':<{max_container_width}}")

    ps_output = subprocess.run(
        ["ps", "-e", "h", "-o", "pid,pcpu,pmem,comm", "--sort=-pcpu"],
        capture_output=True, text=True
    ).stdout.strip()

    if not os.path.exists("/usr/bin/incus"):
        print("Incus is not installed. Not reporting Instance load averages.")
        return

    incus_list = subprocess.run(
        ["/usr/bin/incus", "list", "--format=json"],
        capture_output=True, text=True
    )

    try:
        containers = json.loads(incus_list.stdout)
    except json.JSONDecodeError:
        print("Error: Failed to parse JSON from 'incus list'. Is Incus running and accessible?")
        return

    container_pid_map = {
        str(container.get("state", {}).get("pid")): container["name"]
        for container in containers if container.get("status") == "Running"
    }

    for line in ps_output.split("\n"):
        pid, cpu, mem, command = line.split(maxsplit=3)
        if float(cpu) <= 0.3:
            continue
        container = "host"

        process_ns = f"/proc/{pid}/ns/pid"
        if os.path.exists(process_ns):
            process_ns_link = os.readlink(process_ns)
            for host_pid, container_name in container_pid_map.items():
                host_ns = f"/proc/{host_pid}/ns/pid"
                if os.path.exists(host_ns) and os.readlink(host_ns) == process_ns_link:
                    container = container_name
                    break

        print(f"{pid:<10}{cpu:<6}{mem:<6}{command:<23}{container:<{max_container_width}}")

def main():
    THRESHOLD = 3
    uname = os.uname().nodename

    if not os.isatty(1):
        uptime_output = get_uptime()
        load_avg = float(uptime_output.split("load average:")[1].split(",")[0].strip())
        if load_avg < THRESHOLD:
            return
        print(subprocess.run(["date"], capture_output=True, text=True).stdout.strip())

    print("NODE:")
    print(uname, get_uptime())
    print()

    if not os.path.exists("/usr/bin/incus"):
        print("Incus is not installed. Not reporting Instance load averages.")
        return

    incus_list = subprocess.run(["/usr/bin/incus", "list", "--format=json"], capture_output=True, text=True)
    try:
        containers = json.loads(incus_list.stdout)
    except json.JSONDecodeError:
        print("Error: Failed to parse JSON from 'incus list'. Is Incus running and accessible?")
        return

    # Calculate the maximum container name length
    running_containers = [c["name"] for c in containers if c.get("status") == "Running"]
    max_container_width = max(len(name) for name in running_containers) + 1 if running_containers else 15
    max_container_width = max(max_container_width, len("CONTAINER"))  # Ensure header fits

    # Print container info with dynamic width
    print(f"{'CONTAINER':<{max_container_width}}{'UPTIME':<18}{'TIME':<12}{'USERS':<11}{'LOAD AVERAGE':<16}")
    for c in containers:
        if c.get("status") == "Running":
            container_name = c["name"]

            uptime_output = subprocess.run(
                ["/usr/bin/incus", "exec", container_name, "--", "cat", "/proc/uptime"],
                capture_output=True, text=True
            ).stdout.strip()
            try:
                container_uptime_seconds = float(uptime_output.split()[0])
                days = int(container_uptime_seconds // 86400)
                hours = int((container_uptime_seconds % 86400) // 3600)
                minutes = int((container_uptime_seconds % 3600) // 60)
                uptime_formatted = format_time(days, hours, minutes)
            except Exception:
                uptime_formatted = "N/A"

            uptime_cmd_output = subprocess.run(
                ["/usr/bin/incus", "exec", container_name, "--", "uptime"],
                capture_output=True, text=True
            ).stdout.strip()
            try:
                users = next((s.split()[0] for s in uptime_cmd_output.split(",") if "user" in s), "0")
                load_avg = uptime_cmd_output.split("load average:")[-1].strip()
            except Exception:
                users = "N/A"
                load_avg = "N/A"

            user_text = f"{users} user" if users == "1" else f"{users} users"
            print(f"{container_name:<{max_container_width}}{uptime_formatted:<18}{time.strftime('%H:%M:%S'):<12}{user_text:<11}{load_avg:<16}")

    print()
    map_processes_to_containers(max_container_width)

if __name__ == "__main__":
    main()

#
# Copyright (c) 2008-2025 Michael Stauber,  SOLARSPEED.NET
# Copyright (c) 2008-2025 Greg Kuhnert,     COMPASSNETWORKS.COM.AU
# Copyright (c) 2008-2025 Team BlueOnyx,    BLUEONYX.IT
# All Rights Reserved.
# 
# 1. Redistributions of source code must retain the above copyright 
#    notice, this list of conditions and the following disclaimer.
# 
# 2. Redistributions in binary form must reproduce the above copyright 
#    notice, this list of conditions and the following disclaimer in 
#    the documentation and/or other materials provided with the 
#    distribution.
# 
# 3. Neither the name of the copyright holder nor the names of its 
#    contributors may be used to endorse or promote products derived 
#    from this software without specific prior written permission.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN 
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
# POSSIBILITY OF SUCH DAMAGE.
# 
# You acknowledge that this software is not designed or intended for 
# use in the design, construction, operation or maintenance of any 
# nuclear facility.
#
